User-defined operators for UDVTs

cameel opened the pull request:
Fixes #11969. Part of #13718.
~Depends on #13741.~ Merged.

**Here is a high-level description of the changes introduced by the PR: https://github.com/ethereum/solidity/issues/13718#issuecomment-1341058649. I recommend reading that before reviewing.**

This PR replaces #12362, which has accumulated too many comments and will be hard to get through to anyone wanting to do a fresh review. With the simplifications, fixes and extra tests, I think the code has diverged enough from the original that it will be easier to look at on its own. The code has also been trimmed down quite a bit, with a series of refactor PRs merged separately and support for structs moved to a separate branch ([`user-defined-operators-for-structs`](https://github.com/ethereum/solidity/tree/user-defined-operators-for-structs)).

I'm going to close the original PR by @wechman but before that I'll restore it to the state it was in before I took it over.

wechman committed:
User-defined operators: Code generation

wechman committed:
User-defined operators: Analysis

wechman committed:
User-defined operators: Parsing and grammar

wechman committed:
User-defined operators: AST

wechman committed:
User-defined operators: Tests

wechman committed:
User-defined operators: Documentation

cameel committed:
Fix typo in test name: singed -> signed

nikola-matic commented on docs/contracts/using-for.rst:
```diff
@@ -150,3 +167,38 @@ if you pass memory or value types, a copy will be performed, even in case of the
+        require(result32 <= type(uint16).max, "Divide overflow");
+        return UFixed16x2.wrap(uint16(a32 * SCALE / b32));
+    }
+
```
```suggestion
```
Should there be a double newline here?

cameel replied:
I don't think we have a rule about this but looks superfluous to me. Removed.

nikola-matic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+	_operation.annotation().userDefinedFunction = operatorDefinition;
+
+	TypePointers const& returnParameterTypes = _operation.userDefinedFunctionType()->returnParameterTypes();
+	if (operatorDefinition && returnParameterTypes.size() >= 1)
```
```suggestion
	if (operatorDefinition && !returnParameterTypes.empty())
```

nikola-matic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1763,31 +1813,106 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+				)
+			);
+
+		if (returnParameterTypes.size() >= 1)
```
```suggestion
		if (!returnParameterTypes.empty())
```

nikola-matic commented on libsolidity/analysis/ControlFlowBuilder.cpp:
```diff
@@ -57,24 +57,62 @@ unique_ptr<FunctionFlow> ControlFlowBuilder::createFunctionFlow(
+				m_currentNode = nextNode;
+				return false;
+			}
+
```
```suggestion
```

nikola-matic commented on libsolidity/ast/ASTJsonImporter.cpp:
```diff
@@ -383,15 +384,32 @@ ASTPointer<InheritanceSpecifier> ASTJsonImporter::createInheritanceSpecifier(Jso
+		astAssert(!_node["libraryName"].isArray());
+		astAssert(!_node["libraryName"]["operator"]);
 		functions.emplace_back(createIdentifierPath(_node["libraryName"]));
+		operators.emplace_back();
```
What are we emplacing back here? I'm assuming `nullopt`, but it would be nice to be explicit if so.

cameel replied:
Yeah, `optional<Token>()` so it's `nullopt`. Parameters are forwarded to the constructor and for `optional` they're optional.

nikola-matic commented on libsolidity/ast/ASTJsonImporter.cpp:
```diff
@@ -383,15 +384,32 @@ ASTPointer<InheritanceSpecifier> ASTJsonImporter::createInheritanceSpecifier(Jso
+			{
+				Token const token = scanSingleToken(function["operator"]);
+				astAssert(util::contains(frontend::userDefinableOperators, token));
+				operators.emplace_back(optional<Token>{token});
```
```suggestion
				operators.emplace_back(token);
```

nikola-matic commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -672,6 +672,41 @@ void IRGeneratorForStatements::endVisit(Return const& _return)
+
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(
+			*_unaryOperation.annotation().type == *function->returnParameters().at(0)->type(),
```
```suggestion
			*_unaryOperation.annotation().type == *function->returnParameters()[0]->type(),
```
or `front()`.

nikola-matic commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -775,6 +810,43 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
+
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(
+			*_binOp.annotation().type == *function->returnParameters().at(0)->type(),
```
```suggestion
			*_binOp.annotation().type == *function->returnParameters()[0]->type(),
```

nikola-matic commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -972,19 +973,43 @@ ASTPointer<UsingForDirective> Parser::parseUsingDirective()
+				advance();
+			}
+			else
+				operators.emplace_back();
```
I'm assuming the empty emplace will default initialize an `optional` - so `nullopt` here, but I'm not entirely sure why this is done? I know you've mentioned in a [comment](https://github.com/ethereum/solidity/pull/13790/files#diff-3f7b7120872e5e747b692c9a8bd8310aaf6e4ec285ef300ec4d4d32617fb5f8aR703) that the size of operators and functions must match - does the same apply here?
In any case, if this is in fact the case, I would prefer for calls such as this to be `emplace_back(nullopt)`.

cameel replied:
> I know you've mentioned in a [comment](https://github.com/ethereum/solidity/pull/13790/files#diff-3f7b7120872e5e747b692c9a8bd8310aaf6e4ec285ef300ec4d4d32617fb5f8aR703) that the size of operators and functions must match - does the same apply here?

Yes. We have two parallel vectors. One has functions and the other operators they implement (if the operator is `null`, the function is attached to the type instead). If we skip `null` values, you won't necessarily find the operator corresponding to n-th function at index `n`.

Technically, we could safely skip the final empty values in the `operators` array but having them there makes things more regular.

nikola-matic commented on test/libsolidity/semanticTests/operators/userDefined/custom_fixed_point_type_with_operators.sol:
```diff
@@ -0,0 +1,22 @@
+
+contract C {
+    function applyInterest(Fixed value, Fixed percentage) public pure returns (Fixed result) {
+        return value + value * percentage;
```
Attaching a function to an operator will be evaluated in the order of precedence of the operator, right? I.e. `1 + 2 * 3` will be `7` and not `9`. I've not yet gone over all of the tests, but it would be nice if we had a more trivial example of this in action in the form of a test, i.e. just some simple arithmetic.

cameel replied:
Yeah, the precedence rules do not change. The order implicitly comes from the order in which AST nodes are visited. The PR does not change that order.

But good point. The original PR in general had almost no tests checking that behavior in more complex expressions is not changed. I'll add some.

cameel replied:
Done. Added `calling_all_operators_precedence.sol` that makes sure that the precedence matches the built-in operators.

nikola-matic commented on test/libsolidity/syntaxTests/controlFlow/userDefinedOperators/binary_operator_udvt.sol:
```diff
@@ -0,0 +1,17 @@
+contract C {
+    function f() public pure {
+        X.wrap(1) + X.wrap(Y.unwrap(Y.wrap(2) + Y.wrap(3)));
+        X.wrap(4) + X.wrap(5); // Unreachable
```
~~Can you explain why this is unreachable?~~ Ah, didn't see the revert :)

cameel commented:
Added a change that disallows non-pure suffixes. Originally I wanted it to be a separate PR but this turned out to be simple enough that I think it's fine to add it here. It being a separate commit should be enough.

cameel commented on docs/grammar/SolidityParser.g4:
```diff
@@ -311,11 +311,35 @@ errorDefinition:
+	| LessThan
+	| LessThanOrEqual
+	| NotEqual
+	| Not;
```
`!` currently returns the same type it takes. I actually found it a bit awkward recently when I wanted to implement it for a type based on `uint`. I wonder if it wouldn't make more sense for it to return `bool` instead.

ekpyron replied:
Yes, it should be ``bool``. ``~`` is enough for a negationy thing within the type.

cameel replied:
Changed to `bool`.

cameel commented on docs/grammar/SolidityParser.g4:
```diff
@@ -311,11 +311,35 @@ errorDefinition:
+	| BitNot
+	| BitOr
+	| BitXor
+	| Sar
+	| Shl
+	| Add
+	| Div
+	| Exp
```
Built-in versions of `<<`, `>>` and `**` require an unsigned operand on the right side. This is another thing that was a bit awkward when I tried to implement them for a type that's just a straight wrapper over `int`:
```solidity
type T is int;
using {exp as **} for T;

function exp(T a, T b) pure returns (T) {
    require(T.unwrap(b) >=0);
    return T.wrap(T.unwrap(a) ** uint(T.unwrap(b)));
}
```

Note that it requires a cast and a check for values that are out of range. Maybe we should allow the types to be different in this case?

Or maybe we should disallow them for now? Note that these operators were not user-definable in the initial implementation but the restriction was relaxed in review: https://github.com/ethereum/solidity/pull/12362#discussion_r953894665.

EDIT: Mixed up `uint` and `int`. The right argument actually has to be unsigned, not signed.

ekpyron replied:
Exponentiation should definitely take an unsigned integer as operand, i.e. be ``'a -> uint -> 'a``.
The dogma is not to have same types everywhere, the dogma is to have only one polymorphic type occuring in the generic signature of the operator (which is what any concrete implementation for a concrete type implements). I.e. for *any* exponentiation at *any* type the concrete type `uint` and no other type can be the exponent type.

For shifts it's a bit trickier, let me think a moment there...

ekpyron replied:
Similarly to that (I hope at least :-)?) comparisons are not ``'a -> 'a -> 'a``, but fixed ``'a -> 'a -> bool``.

cameel replied:
Well, if it makes sense to use `uint` for exponentiation then for shifts probably too. They're like division/multiplication by `2^n` and the parameter `n` is an exponent too.

cameel replied:
So now the questions is - should it be specifically `uint` or any unsigned type?

cameel replied:
> Similarly to that (I hope at least :-)?) comparisons are not `'a -> 'a -> 'a`, but fixed `'a -> 'a -> bool`.

Yeah, comparisons are `'a -> 'a -> bool`.

Like I said, the original PR did not have these operators and they were added later - without much consideration for parameter types. I wasn't sure if it was a good idea ito add n the first place because of this ambiguity, but the way you describe it now with `uint`s makes sense to me.

ekpyron replied:
Hm, yeah, that's indeed the core question... conceptually it has to be a single fixed concrete type... but *also* all builtin exponentiations should be representable as user-defined operators eventually... which kind of means it has to be ``uint256`` specifically... but then implicit conversions for operands come into play...

It's no coincidence that we have weird cases that produce warnings for exponent types generally :-)...

The safest course of action would probably be to postpone all three operators and not rush design decisions on this now.

ekpyron replied:
Ultimately it will *probably* be `uint256`. But yeah, maybe better to think through implicit conversions on operands and relate that to the builtin versions of those operators before rushing.

cameel replied:
ok then. I'll remove them and we can consider that properly for 0.8.19.

cameel replied:
`<<`, `>>` and `**` disallowed.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = operandType->unaryOperatorResult(op);
+	set<FunctionDefinition const*> matchingDefinitions = operandType->operatorDefinitions(
+		_operation.getOperator(),
```
```suggestion
		op,
```

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1763,31 +1813,106 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-			(!result.message().empty() ? " " + result.message() : "")
+		SecondarySourceLocation secondaryLocation;
+		for (FunctionDefinition const* definition: matchingDefinitions)
+			secondaryLocation.append("Candidate definition:", definition->location());
```
This loop could be moved to inside the next `if` block, like it was done in `visit(UnaryOperation)`. I think it makes more sense that way.

0xalpharush commented on test/libsolidity/semanticTests/operators/userDefined/calling_all_operators.sol:
```diff
@@ -0,0 +1,110 @@
+    function test_gt(int128 x) public pure returns (bool) { return w(x) > w(2); }
+    function test_leq(int128 x) public pure returns (bool) { return w(x) <= w(2); }
+    function test_geq(int128 x) public pure returns (bool) { return w(x) >= w(2); }
+    function test_not() public pure returns (bool) { return !w(1); }
```
~I'm confused about the semantics given that this implemetation returns a boolean yet should return `Int` based on the statement, "! returns the same type it takes rather than bool. Should it be returning bool instead?", [here](https://forum.soliditylang.org/t/feature-preview-user-defined-operators/1435).~

 **EDIT** I see this behavior was changed with https://github.com/ethereum/solidity/pull/13790/commits/271b6f46fc03b0e3c0c9291aca31a18bf8b0341e

Also, I'd expect the "not" operator to only apply to user defined types whose underlying type can also use the operator. In this case, I would expect a similar error to trying `!int128(1)`: `error[4907]: TypeError: Unary operator ! cannot be applied to type int128.`

0xalpharush replied:
~I tried this locally (with a binary from the forum post) and observed the behavior explained in the feature preview, so I'm not sure why this is passing?~
```
Error: Wrong return parameters in operator definition. The function "not" needs to return exactly one value of type Int to be used for the operator !.
  --> test.sol:63:34:
   |
63 | function not(Int x) pure returns (bool) {
   |                                  ^^^^
```
My concern regarding the second point still stands

cameel replied:
There's some confusion here so let me explain what happened:
- I posted the thread. At that point `!` was still returning the argument type. I wasn't sure if that's a good design so I included a question about it.
- I also posted this question here (https://github.com/ethereum/solidity/pull/13790#discussion_r1072798198). Daniel answered with a good point that it should be returning `bool`.
- Later I applied this change and pushed the change here. Now in the PR you have it returning `bool` but the thread still has the binary from before that change.
- Another change here compared to the binaries in the thread is that `<<`, `>>` and `**` can no longer be redefined (see https://github.com/ethereum/solidity/pull/13790#discussion_r1072807608 for discussion).
- I'm going to update the binary in the thread and say what changed, either today or on Monday. I want to keep it up to date but I also think that posting there after every single change in the PR would be both tedious for me and confusing to people so I'm going to do it only after a few changes accumulate. Hopefully there won't be any more changes at this point anyway. For the purpose of high-level discussion and user feedback I think that should be enough granularity.
- I did not expect someone to actually come to the PR and review it (and thanks for looking at it! - feel free to comment on the code). If you're looking at the PR directly it's best to use the current binary from the CI. The ones on the forum link to what was in CI a few runs ago. You can get to the current runs via the "Details" links on CI checks below. The jobs that give you static binaries are : `b_ubu_static`, `b_win`, `b_osx` and `b_ems`.

cameel replied:
> Also, I'd expect the "not" operator to only apply to user defined types whose underlying type can also use the operator. In this case, I would expect a similar error to trying `!int128(1)`: `error[4907]: TypeError: Unary operator ! cannot be applied to type int128.`

That's actually what happens, You can only use these operators on user-defined types. This test may be confusing because for every operator definition it also has a `test_xyz()` helper function that calls it and these functions take raw integers and wrap them in the custom `Int` type just before using the operator. The operations inside them are actually performed on the custom type, not on `int128`. This might not be apparent because the test hides the wrapping in a another helper function: `w()`.

We use this style with test helpers because in the test expectations we can only call external functions. Operators are internal and cannot be used in expectations directly.

I think I can actually rewrite this test to be more straightforward. For exampie, we could have the helpers use `Int` rather than `int128`. We could also group the operations in one helper. I'll push a change with that in a moment.

cameel replied:
Please check out the updated test. I hope it's clearer now.

0xalpharush replied:
Thanks for the clarification and pointers. That's my for bad using an outdated binary and reviewing recent changes. I was able to download an up-to-date binary.

My point was related to code like the following:
```
type Int is int128;
using {
    not as !
} for Int;

function not(Int x) pure returns (bool) {
    return Int.unwrap(x) == 0;
}

function a() pure returns (bool) {
      require(!Int.wrap(1));
      if (!Int.wrap(1)) {
          return false;
      }
      return !Int.wrap(1);
  }
// function does_not_compile() pure returns (bool) {
//       require(!int128(1));
//       if (!int128(1)) {
//           return false;
//       }
//       return !int128(1);
//   }
```
If I were to try this with the underlying type, `int128`, I would receive a type error indicating that the unary not operator is not available for this type. I don't see why wrapping it in a type alias should change that behavior. What the function `not` is doing is arbitrary and isn't related to logical negation as that operation cannot be performed on it's underlying type. It would make since to instead define a helper function instead:
```
function is_zero(Int x) pure returns (bool) {
     return Int.unwrap(x) == 0;
}
function b() pure returns (bool) {
      require(is_zero(Int.wrap(1)));
      if (is_zero(Int.wrap(1))) {
          return false;
      }
      return is_zero(Int.wrap(1));
}
```
My suggestion would be to only allow the not operator for user defined types whose underlying type is `bool`. Otherwise, the not operator will just be used to check an arbitrary condition on a value type and no longer mean logical negation. I think enforcing the latter option would be best for clarity and be in line with the existing semantics of the not operator.

cameel replied:
Ah, now I see what you meant. You make a good point. I'm still on the fence with that though so let's wait for some input from @ekpyron or @chriseth. I think you should also raise that point in the forum thread so other users can chime in. That's exactly the kind of a design issue I was hoping for people to point out and state their opinion on before we release it.

If I had to argue for keeping it as is, I'd say that the underlying type of the UDVT should be treated as an implementation detail. You might want to implement a type with some boolean-like behavior but backed by something other than `bool`. So in that sense allowing `!` is ok because there are UDVTs for which it semantically makes sense. And we can't really prevent people from doing all unreasonable things with UDVTs in all cases. You could define operators with completely crazy semantics and we can't stop you :) This is a trade-off and the question is if restricting it does more good than harm.

On the other hand, one thing that reasonates with me in your argument is that I'm pretty sure people will use this feature to implement truthy/falsy expressions, like `!x` instead of `x == 0`. I think this style makes conditions less readable and is also error prone (`0` is false but sometimes gets forgotten when your check is meant to catch `null`). Still, that's just my personal opinion and there's no `null` in Solidity so there's much less danger in doing that.

Another good argument for disallowing it would be that we did just that with other operators where the design has some uncertainties, like `**`, `++` or `&&`. This way we can still enable them later in a non-breaking release when we have a better idea how it will mesh with other features we have in the pipeline now (especially generics). Allowing them now and disallowing later would be breaking.

cameel replied:
In the end we decided to disallow `!` in the initial release.

ekpyron commented on docs/contracts/using-for.rst:
```diff
@@ -20,7 +22,15 @@ The first part, ``A``, can be one of:
-  all non-private functions of the library are attached to the type.
+  all non-private functions of the library are attached to the type
+  as member functions
+- a list of assignments of pure file-level or public/internal/private library functions to operators
```
This kind of suggests that it can *either* be a list of functions assigned to operators, *or* a list of member functions, while a combination of both seems possible as well (``attaching_and_defining_operator_with_same_function.sol`` is the only test case for that, but it is a test case for that)

By the way: how does the case in that test case occur in the ast? With the same function name twice once with an additional ``operator`` field, once without?

cameel replied:
Very good point. I have rewritten this bit to make it clearer. I also added a new test case explicitly checking that you can mix defining operators and attaching functions.

cameel replied:
> By the way: how does the case in that test case occur in the ast? With the same function name twice once with an additional `operator` field, once without?

Yes, exactly:

```solidity
using {add as +, add} for Int;
```
```json
            "functionList":
            [
                {
                    "function":
                    {
                        "id": 3,
                        "name": "add",
                        "nameLocations":
                        [
                            "27:3:0"
                        ],
                        "nodeType": "IdentifierPath",
                        "referencedDeclaration": 33,
                        "src": "27:3:0"
                    },
                    "operator": "+"
                },
                {
                    "function":
                    {
                        "id": 4,
                        "name": "add",
                        "nameLocations":
                        [
                            "37:3:0"
                        ],
                        "nodeType": "IdentifierPath",
                        "referencedDeclaration": 33,
                        "src": "37:3:0"
                    }
                }
            ],
```

ekpyron replied:
Yeah, doing it like this in the AST of course has the danger of resulting in silent weird behaviour for tooling that isn't aware of the new field... but not sure there's a good way to avoid that and we did ping tooling about this...

ekpyron replied:
Making it ``{ "operator": { "function": { /* as now */ }, "operation": "+" }}`` as entries in ``functionList`` (which is kind of a misnomer now anyways, but which we can't change anyways) would be an alternative without that problem, but not sure if that's actually better or worse...

cameel replied:
Changed to `{"operator": "+", "definition": { /* as now */ }}` for operators.

ekpyron commented on docs/contracts/using-for.rst:
```diff
@@ -38,6 +48,13 @@ then the type (``uint``) has to be implicitly convertible to the
 
+If you define an operator for a user-defined type (``using {f as +} for T``), then
+the type (``T``), types of function parameters and the type of the function return value
+have to be the same. The type (``T``) does not include data location.
+But, data location of the function parameters and function return value must be
+the same. One exception from this is the result type of comparison and boolean operators
```
This suggests that there can be data locations involved, even though we restricted to value-types.

ekpyron replied:
I'd generally try not to phrase this as "usually it's all the same type, but there are exceptions", but that each operator requires types of a specific generic signature and then list those signatures for all supported operators...

cameel replied:
Ah, good point. I missed that when extracting support for structs.

> I'd generally try not to phrase this as "usually it's all the same type, but there are exceptions" 

Yeah, I'll rephrase it.

BTW, These docs are what was originally in the PR and I left it mostly untouched but I'm not very satisfied with it. I'm planning to rewrite the whole section on `using for` once I'm done with suffixes. I wanted to leave it for later because I think the whole section could be clearer, not just the bits that were added for operators.

cameel replied:
Done. I added a table listing possible signatures. It's a bit verbose and repetitive so not sure how well it will work when we add more operators to it. Still I kinda like it this way. Please take a look and tell me what you think.

ekpyron replied:
Eventually, the operators will actually have generic in-language definitions, then we can just link the standard library here instead. The verbose part is partly lack of https://github.com/ethereum/solidity/issues/1349 and lack of fixed syntax conventions for type variables :-). But lacking those, the current list is good!

cameel replied:
> partly lack of #1349

Oh, I wasn't even aware of that issue. Should I change to a notation that resembles this?

ekpyron replied:
No, I just meant in the future we can make it mire concise, but for now let's stick to actually valid syntax :-)

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3859,6 +3986,95 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
+					"Only pure functions can be used to define operators."
+				);
+
+			bool identicalFirstTwoParameters = (parameterCount < 2 || *parameterTypes.at(0) == *parameterTypes.at(1));
```
If I was @cameel, I might complain about something that has none or only one parameter not really having identical first two parameters ;-).

cameel replied:
That's because you're looking at it in absolute terms :) I think it's still a big relative improvement over the [monster conditions in the old PR](https://github.com/ethereum/solidity/pull/12362/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R3965-R3986) :)

cameel replied:
Also, when you have less than two parameters, it's not the case where the first two parameters don't match, so I find this way of formulating it quite intuitive.

ekpyron replied:
Arguing as a Russellian, it's true, that it's false that the first two parameters don't match in this case, but it's also true that it's false that they do match ;-P.

ekpyron replied:
But yeah, the checking here generally seems a bit convoluted, but we can also keep it.

cameel replied:
Yeah, there's just a lot going on in these conditions and I did not find a way to do it better. I think it's at least better than the original. This seemed like a good compromise between being precise and keeping it readable.

ekpyron commented on docs/contracts/using-for.rst:
```diff
@@ -37,6 +41,50 @@ If you use a list of functions (e.g. ``using {f, g, h, L.t} for uint;``),
+The following operators can be defined this way:
+
++------------+----------+---------------------------------------------+
+| Category   | Operator | Possible signatures                         |
```
```suggestion
| Category   | Operator | Signature                                   |
```

It's the one and only signature of such a function for any given type ``T``, isn't it :-)?

cameel replied:
There are two possible signatures for `-`.

ekpyron replied:
Hm, I'd still think of those as two distinct operators (they're distinguished at the parsing/grammar level fwiw) each with unique signature, but yeah, fair enough, that makes it a bit fuzzy

cameel replied:
The problem for me is that having two different operators with the same name makes it hard to clearly refer to them in the docs. I find it easier to treat them as the same operator in docs and just say it has multiple signatures or variants. In the end it's just a different way to refer to the same thing anyway.

ekpyron replied:
Well, not important enough to hold this up anyways. Even though I'd say it's not the same thing you refer to :-). But let's just keep it as is anyways for now.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -410,6 +410,38 @@ bool ExpressionCompiler::visit(TupleExpression const& _tuple)
+		FunctionType const* functionType = _unaryOperation.userDefinedFunctionType();
+		solAssert(functionType);
+
+		functionType = dynamic_cast<FunctionType const&>(*functionType).withBoundFirstArgument();
```
Why indirect this via a function with a bound first argument? I don't see any first argument being bound in any of this - I'd just call the function with the operators as arguments, is there any difference?

cameel replied:
Not sure. It seems to have been like this in @chriseth's initial implementation: 3bd047f1881073b2002a31baed7fb974dbc20c57. Unless that was fixed up into it later but I doubt it. In any case it's how it was in the original PR and @chriseth did not complain about it even though he paid a lot of attention to the codegen part (e.g. https://github.com/ethereum/solidity/pull/12362#discussion_r915822159). I suspect this was started as a copy-paste of the code for handling bound functions and this part was not wrong so it survived in that form even if it was not strictly necessary.

But yeah, now that you pointed it out, it seems unnecessary other than to access `selfType()` which is also just a helper to access the first argument type. I will switch to a plain function type.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -502,7 +534,39 @@ bool ExpressionCompiler::visit(BinaryOperation const& _binaryOperation)
+
+		FunctionType const* functionType = _binaryOperation.userDefinedFunctionType();
+		solAssert(functionType);
+		functionType = dynamic_cast<FunctionType const&>(*functionType).withBoundFirstArgument();
```
Same here. Even though we reuse the `using` syntax, operators are not really *bound* in any way, let alone in one of two operand arguments... same for via-IR...

cameel committed:
Warn about missing user-defined operator support in SMTChecker

cameel commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -449,11 +449,15 @@ void SMTEncoder::endVisit(UnaryOperation const& _op)
+	if (TokenTraits::isBitOp(_op.getOperator()) && !*_op.annotation().userDefinedFunction)
 		return bitwiseNotOperation(_op);
 
 	createExpr(_op);
 
+	// User-defined operators are essentially function calls.
+	if (*_op.annotation().userDefinedFunction)
+		return;
```
@leonardoalt Is this the right way to handle user-defined operators? I suspect not since my test are not detecting assertion violations as they do when I call operator functions directly, but `createExpr(Expression)` is basically what seems to happen for an internal function call.

ekpyron replied:
We can always just emit a warning about user-defined operators not being supported by the smtchecker yet and take it from there. Not sure it's a good idea to try to properly implement them there in this PR.

leonardoalt replied:
Agree with @ekpyron

leonardoalt replied:
> but `createExpr(Expression)` is basically what seems to happen for an internal function call.

This is only in the SMTEncoder side though. BMC and CHC handle control flow separately, so maybe some extra support is needed there to interpret them as function calls.

cameel replied:
Warning added:

```
User-defined operators are not yet supported by SMTChecker. This invocation of operator + has been ignored, which may lead to incorrect results.
```

Other than that I left the things I added as is - since there wasn't that much of it anyway. Just a few `if`s preventing user-defined operators being processed as normal operators even with the warning. Do you think it can stay that way or should I actually get rid of the `createExpr(_op)` part when it's a custom operator?

cameel commented on test/libsolidity/smtCheckerTests/operators/userDefined/user_defined_operator_matches_equivalent_function_call.sol:
```diff
@@ -0,0 +1,77 @@
+        assert(x - y == sub(x, y)); // FIXME: should hold
+        assert(-x == unsub(x)); // FIXME: should hold
+        assert(x * y == mul(x, y)); // FIXME: should hold
+        //assert(x / y == div(x, y)); // FIXME: ICE due to https://github.com/ethereum/solidity/issues/13900
```
This seems to be a bug unrelated to the PR so I reported it separately. Does not seem critical to fix but it does get in the way of some tests.

cameel replied:
The underlying bug was fixed so uncommented this now.

ekpyron commented on libsolidity/analysis/ControlFlowBuilder.cpp:
```diff
@@ -57,24 +57,61 @@ unique_ptr<FunctionFlow> ControlFlowBuilder::createFunctionFlow(
 		case Token::Or:
 		case Token::And:
 		{
 			visitNode(_operation);
```
Wait - is this visited twice now? Is that intentional?

ekpyron replied:
Not that it matters too much, but a bit weird - the second one here can be removed, can't it? The third one below as well...

ekpyron replied:
``ASTConstVisitor::visit(_operation);`` then does it yet another time...

ekpyron replied:
But well, it's not like it hurts to call it multiple times - and even if the source locations ended up incorrect, it'll still only result in weird error messages and not in actually broken behaviour, so fine either way.

cameel replied:
Need to take a closer look, but in the original PR was actually skipping some `visitNode()` calls that should have been there. Need to check if it's that or if not there's too many of them :)

cameel replied:
Looks like it was moved in and out of the switch in the original PR. There is even a comment about that (https://github.com/ethereum/solidity/pull/12362/files#r931895231). Something must have gone wrong with that because apparently it was added inside but not removed outside. I must have missed that in review too. There were actually a few changes like this that ended up marked as resolved but not applied and I had to ask about them.

In any case, I'm removing it.

cameel commented on test/libsolidity/smtCheckerTests/operators/userDefined/user_defined_operations_on_constants_fail.sol:
```diff
@@ -0,0 +1,104 @@
+// Warning 6756: (2062-2073): User-defined operators are not yet supported by SMTChecker. This invocation of operator & has been ignored, which may lead to incorrect results.
+// Warning 6756: (2062-2081): User-defined operators are not yet supported by SMTChecker. This invocation of operator == has been ignored, which may lead to incorrect results.
+// Warning 6756: (2114-2123): User-defined operators are not yet supported by SMTChecker. This invocation of operator ^ has been ignored, which may lead to incorrect results.
+// Warning 6756: (2114-2131): User-defined operators are not yet supported by SMTChecker. This invocation of operator == has been ignored, which may lead to incorrect results.
+// Warning 6156: (2164-2168): User-defined operators are not yet supported by SMTChecker. This invocation of operator ~ has been ignored, which may lead to incorrect results.
```
Note that there are two different error codes for this (oddly, `error_codes.py` gave them misleadingly similar codes - WTF?). From tooling perspective I think it's not great to have multiple codes for the same thing but our current error check script is a bit annoying about this. It won't let you use the same code twice even if you know it's the same thing.

I considered extracting the piece reporting the warning into a helper but maybe it's an overkill for an error that will probably won't live longer than one release and won't be seen too often...

ekpyron replied:
Yeah, it's fine - indeed a weird coincidence though :-).

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+		if (matchingDefinitions.size() >= 2)
+		{
+			SecondarySourceLocation secondaryLocation;
+			for (FunctionDefinition const* definition: matchingDefinitions)
+				secondaryLocation.append("Candidate definition:", definition->location());
```
Oh wow, this has nondeterministic order, doesn't it? Same in the binary case. Should be fixed by choosing e.g. an AST-ID ordering.

cameel replied:
True. Good catch. Should be fixed now (did not build locally for me yet but CI will catch it if it's not).

I tried to go for AST ID ordering at first but then I'd have to change some includes from `ASTForward.h` to `AST.h` because `ASTNode::CompareByID` is nested and can't be forward declared. I guess a `vector` is fine here. Fortunately we don't really need to look up stuff in it. We only check how many definitions it has and sometimes enumerate them.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = operandType->unaryOperatorResult(op);
+	set<FunctionDefinition const*> matchingDefinitions = operandType->operatorDefinitions(
```
i.e. easiest quickest fix for https://github.com/ethereum/solidity/pull/13790/files#r1091037315 would probably be using AST-ID ordering for this set.

cameel replied:
Same here. Changed to `vector`.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+
+	// By default use the type we'd expect from correct code. This way we can continue analysis
+	// of other expressions in a sensible way in case of a non-fatal error.
+	Type const* resultType = operandType;
```
For the record: Technically, the operand type is not always the type we'd expect from correct code :-). Let's leave it like this anyways - improving this will become much easier once we have a first type class/trait system that will solve all this generically by matching against global polymorphic signatures and with which all of this will just become a special case. And it's only error conditions affected anyways.

cameel replied:
Well, I guess I was thinking mostly about the user-defined case here and there the result type has to match the operand type, right?

Do you mean that built-in stuff like, e.g. `!` returns `bool` rather than operand type?

ekpyron replied:
Ah, sorry, actually for the currently allowed unary ops, it's true - and for binary ones below it's actually more specific.

ekpyron replied:
We just need to be careful when extending this for more operators for which this may no longer be the case :-).

cameel replied:
I mean, I should probably reword this to account for `!`. I did not actually think about this particular example back then.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+			for (FunctionDefinition const* definition: matchingDefinitions)
+				secondaryLocation.append("Candidate definition:", definition->location());
+
+			m_errorReporter.typeError(
```
Also for the record: Technically, we *could* check this at the level of the using statements already and maybe that'd be better - currently, I guess, you can construct valid cases, in which only every actual use of the operator is an error... but yeah, also not worth changing it now.

cameel replied:
Maybe but then we'd have to deal with the using statement not being correct. At this point we can assume that they passed all the checks.

Or... wait, is it possible that we'll visit an operation before all the `using` directives are analyzed? Technically I don't see why it would not be possible but I did not run into that.

cameel replied:
But anyway, I don't think it's possible at `using for` level because there may be multiple directives defining operators for the same type. You have to process all of them to know if the operator definition is unique or not.

ekpyron replied:
Yeah, but we can do that in principle - but yeah, anyways, we won't do it now no matter what.

cameel replied:
Actually, ok, it is possible. I was assuming that `TypeChecker` sets some annotations on `UsingForDirective` but apparently not. It's probably all set by earlier analysis stages so I guess I could detect this while analyzing the directive.

ekpyron replied:
Yeah - but I think we have a lot of analog cases and most of them only throw errors on use - so maybe we just fix that in bulk some time. It's technically breaking for the cases in which we already have this kind of thing.
So yeah, let's keep it as is here for now.

cameel replied:
Is this actually breaking though? As long as good code keeps compiling and bad code keeps having errors, I think this is something we could change.

If you want, I can create an issue to be done after we merge this PR. It could be a good task for a contributor.

While at it, we could also store operator definitions for a type in the AST. It's not very efficient that we rescan all `using for` directives we encounter an operator.

ekpyron replied:
I don't think it's a task for a contributor :-). Checking for all other situations that may be analog alone is hard. But yeah, an issue would be good.
And well, it's technically breaking. There is good code that compiles and then doesn't after the change. The fact that the good code is only good as long as it has redundant statements doesn't change that really :-). But it's a technicality, so we could get away with it anyways.

cameel replied:
Actually, looks like it should not be a problem outside of operators. I.e. attached functions are resolved immediately in `using for` and ambiguous definitions are not accepted.

For operators it's different because you can't tell that just looking at a single `using for`. You have multiple functions defining the same operator, being added by separate directives. I guess that's why the check was not originally in `using for`. But yeah, it's doable so I'll change that. I was going to file an issue for it but since it's only for operators and we postponed the release, I can do it properly already in this PR.

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -775,6 +798,30 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
+		string right = expressionAsType(_binOp.rightExpression(), *function->parameters()[1]->type());
+		solAssert(!left.empty() && !right.empty());
+
+		define(_binOp) <<
+			m_context.enqueueFunctionForCodeGeneration(*function) <<
+			("(" + left + ", " + right + ")\n");
```
This will always generate code for call internally, which is wrong for public library functions.
Same for legacy codegen actually.

Compare

```
type T is uint;

library L {
    function f(T, T) public pure returns (T) {
        assembly { mstore(0x42, 0x42) }
    }
}

using {L.f as +} for T;

contract C {
    function f(uint x, uint y) external returns(uint) {
        T a = T.wrap(x);
        T b = T.wrap(y);
        T c = L.f(a, b);
        return T.unwrap(c);
    }
```

with

```
type T is uint;

library L {
    function f(T, T) public pure returns (T) {
        assembly { mstore(0x42, 0x42) }
    }
}

using {L.f as +} for T;

contract C {
    function f(uint x, uint y) external returns(uint) {
        T a = T.wrap(x);
        T b = T.wrap(y);
        T c = a + b;
        return T.unwrap(c);
    }
```

and e.g. note the delegate call in one, but not the other, in IR output, or the link reference in one, but not the other in ``--bin`` output.

cameel replied:
Damn, you're right. It was already in the original PR but I did miss this in my review.

ekpyron replied:
We can hotfix this by restricting to internal library functions only for now - they make more sense anyways. There won't be enough time for fixing this properly unless we want to postpone the release.

ekpyron replied:
> Damn, you're right. It was already in the original PR but I did miss this in my review.

Yeah, also not sure why I didn't see it before now :-).

cameel committed:
Tests for using enums and fixed-point types as user-defined types

cameel commented on test/libsolidity/syntaxTests/operators/userDefined/using_for_with_operator_at_file_level_for_contract_level_type.sol:
```diff
@@ -0,0 +1,27 @@
+library L {
+    type FixedBytes is bytes1;
+}
+
+contract C {
+    type FixedBytes is bytes2;
+}
+
+interface I {
+    type FixedBytes is bytes3;
+}
+
+function addL(L.FixedBytes, L.FixedBytes) pure returns (L.FixedBytes) {}
+function addC(C.FixedBytes, C.FixedBytes) pure returns (C.FixedBytes) {}
+function addI(I.FixedBytes, I.FixedBytes) pure returns (I.FixedBytes) {}
+
+function unsubL(L.FixedBytes) pure returns (L.FixedBytes) {}
+function unsubC(C.FixedBytes) pure returns (C.FixedBytes) {}
+function unsubI(I.FixedBytes) pure returns (I.FixedBytes) {}
+
+using {addL as +, unsubL as -} for L.FixedBytes global;
+using {addC as +, unsubC as -} for C.FixedBytes global;
+using {addI as +, unsubI as -} for I.FixedBytes global;
+// ----
+// TypeError 4117: (545-600): Can only use "global" with types defined in the same source unit at file level.
+// TypeError 4117: (601-656): Can only use "global" with types defined in the same source unit at file level.
+// TypeError 4117: (657-712): Can only use "global" with types defined in the same source unit at file level.
```
Important consequence of only allowing global operators: `global` is disallowed for types defined inside contracts so it will make it impossible to define operators for them.

cameel commented:
PR ready for review again:
- Disallowed using library functions for operators.
- Disallowed non-global operators.
- Operator bindings are type-checked in `using for` rather than at the point of use.
- More tests for various corner cases.
- Consistently using `TokenTraits::friendlyName()` in error messages.

Also:
- I did not yet implement storing information about operators in a source unit annotation. I'll do that in a follow-up PR not to stall this one.
    - My initial approach didn't work out because I did it in `TypeChecker` and there there's no guarantee that `using for` will be processed before the operator is used. I have another idea but not yet done with it.
- I did not forbid binding the same operator to the same function multiple times. Actually, without having the operators stored in the annotation, it's easier not to forbid it. I might still do it in the PR that adds the annotation.

Commits are a mess right now so I'll be squashing them and pushing here but there should be no more actual change unless requested in the review.

ekpyron commented:
>       * My initial approach didn't work out because I did it in `TypeChecker` and there there's no guarantee that `using for` will be processed before the operator is used. I have another idea but not yet done with it.
> 

Sounds like a job for the ``DeclarationTypeChecker``- the purpose of that existing is exactly to precollect and precheck stuff before type checking to avoid problems like that... but yeah, really no need to do it in this PR, especially if the behaviour is already there.

And we'll need to redesign all of this for https://github.com/ethereum/solidity/issues/13365 eventually anyways - might be wise to start thinking about that more concretely in the foreseeable future actually...

ekpyron commented on Changelog.md:
```diff
@@ -1,6 +1,7 @@
 ### 0.8.19 (unreleased)
 
 Language Features:
+* Allow defining custom operators for user-defined value types via ``using {f as +} for Typename;`` syntax.
```
```suggestion
* Allow defining custom operators for user-defined value types via ``using {f as +} for Typename global;`` syntax.
```

or just dropping the semicolon as

```suggestion
* Allow defining custom operators for user-defined value types via ``using {f as +} for Typename`` syntax.
```

may be better now, right :-)?

cameel replied:
Changed to `using {f as +} for T global`.

ekpyron commented on docs/contracts/using-for.rst:
```diff
@@ -37,6 +42,51 @@ If you use a list of functions (e.g. ``using {f, g, h, L.t} for uint;``),
+|            | ``>=``   | ``function (T, T) pure returns (bool)``     |
++------------+----------+---------------------------------------------+
+
+If an operator supports more than one signature, each one can have a separate definition.
```
```suggestion
Unary and binary ``-`` can have separate definitions.
```
I mean, it's only one operator, no need to be cryptic about it, right :-)?
Fine either way, though.

cameel replied:
I wanted it to be general, and the table already shows `-`, but I guess I can also spell it out given that we only have one such operator. Changed.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -410,6 +410,34 @@ bool ExpressionCompiler::visit(TupleExpression const& _tuple)
+	FunctionDefinition const* function = *_unaryOperation.annotation().userDefinedFunction;
+	if (function)
+	{
+		solAssert(function->isFree() || function->libraryFunction());
```
```suggestion
		solAssert(function->isFree());
```

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -502,7 +530,34 @@ bool ExpressionCompiler::visit(BinaryOperation const& _binaryOperation)
+	FunctionDefinition const* function = *_binaryOperation.annotation().userDefinedFunction;
+	if (function)
+	{
+		solAssert(function->isFree() || function->libraryFunction());
```
```suggestion
		solAssert(function->isFree());
```

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -672,6 +672,30 @@ void IRGeneratorForStatements::endVisit(Return const& _return)
+		setLocation(_unaryOperation);
+
+		solAssert(function->isImplemented());
+		solAssert(function->isFree() || function->libraryFunction());
```
```suggestion
		solAssert(function->isFree());
```

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -775,6 +799,31 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
+		setLocation(_binOp);
+
+		solAssert(function->isImplemented());
+		solAssert(function->isFree() || function->libraryFunction());
```
```suggestion
		solAssert(function->isFree());
```

ekpyron reviewed (commented):
Maybe we should have one or two tests about the evaluation order, so sth like

```
f("A") + f("B");

function f(string memory x) returns (T) {
  emit Event(x);
  return T.wrap(0);
}
```

Apart from that I only had the minor comments above left.

The using-for type checking issue could probably be rather easily solved by filling in annotation in the DeclarationTypeChecker, but we can postpone it.

cameel commented:
> Maybe we should have one or two tests about the evaluation order, so sth like

Added a test that checks evaluation order on some simple cases with binary/unary, multiple operators and operators nested in calls.

cameel commented:
> Sounds like a job for the ``DeclarationTypeChecker``- the purpose of that existing is exactly to precollect and precheck stuff before type checking to avoid problems like that...

I wasn't sure if it would work because I could run into the same problem I have in `TypeChecker`. I.e. it's the `DeclarationTypeChecker` that processes `type T is X` declarations so at the point where I find `using for` the type used in it might not yet have all annotations needed to handle it.

My initial idea was actually to do it in `NameAndTypeResolver`, because it seems like it would be the right place for resolving what the operator name refers to. But I dropped that idea because it runs even earlier than `NameAndTypeResolver` so I have no type info yet.

The approach I wanted to try instead was to scan all `using for` directives once in `TypeChecker` when I'm visiting `SourceUnit` node.

> The using-for type checking issue could probably be rather easily solved by filling in annotation in the DeclarationTypeChecker, but we can postpone it.

Well, it's not complicated to implement but some approaches do not work so needs some experimentation. I was aiming to still get it into this release, just as a separate PR so that it does not  stop the main PR in case it's not ready.

ekpyron commented:
> > Sounds like a job for the `DeclarationTypeChecker`- the purpose of that existing is exactly to precollect and precheck stuff before type checking to avoid problems like that...
> 
> I wasn't sure if it would work because I could run into the same problem I have in `TypeChecker`. I.e. it's the `DeclarationTypeChecker` that processes `type T is X` declarations so at the point where I find `using for` the type used in it might not yet have all annotations needed to handle it.
> 
> My initial idea was actually to do it in `NameAndTypeResolver`, because it seems like it would be the right place for resolving what the operator name refers to. But I dropped that idea because it runs even earlier than `NameAndTypeResolver` so I have no type info yet.
> 
> The approach I wanted to try instead was to scan all `using for` directives in `TypeChecker`, but only once, when I'm visiting `SourceUnit` node.
> 
> > The using-for type checking issue could probably be rather easily solved by filling in annotation in the DeclarationTypeChecker, but we can postpone it.
> 
> Well, it's not complicated to implement but some approaches do not work so needs some experimentation. I was aiming to still get it into this release, just as a separate PR so that it does not stop the main PR in case it's not ready.

The point is not to handle everything elsewhere, but to have one pass collect the statements in appropriate annotations and the next check it.

ekpyron commented on docs/contracts/using-for.rst:
```diff
@@ -7,27 +7,32 @@ Using For
-  only those functions will be attached to the type as member functions.
-  Note that private library functions can only be specified when ``using for`` is inside the library.
+- A list of functions, optionally with an operator name assigned (e.g.
+  ``using {f, g as +, h, L.t} for uint;``).
```
```suggestion
  ``using {f, g as +, h, L.t} for uint``).
```

or 

```suggestion
  ``using {f, g as +, h, L.t} for uint global;``).
```

Or not sure what to do - it's a minor bug we can fix later if need be - may make things a bit confusing to mix everything together here.

cameel replied:
Good point. I'll fix that.

cameel replied:
I now removed the semicolons after such examples.

cameel commented:
I just pushed 4 more commits. Changes only in docs and tests:
- Correction for https://github.com/ethereum/solidity/pull/13790#discussion_r1111967019.
- Syntax test for making a pure external call was actually broken and failing for an unrelated reason. Replaced it with a set of correct syntax tests for external calls and also added one semantic test since that case should actually work. I ran into #13989 while fixing them.
- Removed fixme about #13900 from smt tests.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1763,31 +1800,95 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+		solAssert(parameterTypes.size() == 2);
+		solAssert(parameterTypes[0] && *leftType == *parameterTypes[0]);
+
+		if (*rightType != *parameterTypes[0])
+			m_errorReporter.typeError(
+				5653_error,
+				_operation.location(),
+				fmt::format(
+					"The type of the second operand of this user-defined binary operator {} "
+					"does not match the type of the first operand, which is {}.",
+					TokenTraits::friendlyName(_operation.getOperator()),
+					parameterTypes[0]->humanReadableName()
+				)
+			);
```
~~Why not check this on the using statement as well?~~ Nevermind :-).

cameel replied:
Yeah, we can only check that at the point of use :)

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1763,31 +1800,95 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	TypeResult builtinResult = leftType->binaryOperatorResult(_operation.getOperator(), rightType);
+	set<FunctionDefinition const*, ASTNode::CompareByID> matchingDefinitions = leftType->operatorDefinitions(
+		_operation.getOperator(),
+		*currentDefinitionScope(),
```
Which scope is this exactly? Isn't this the scope that contains the operation itself? Isn't that the wrong scope for the lookup?

cameel replied:
To be perfectly honest, I still don't fully understand how our scopes work but this is the way @chriseth did it originally and I did not find any test case that would break it so I was assuming it's correct.

On the surface, it does seem to make sense. It's the scope where the operator is used so you're only interested in operators visible there.

ekpyron commented on libsolidity/ast/Types.cpp:
```diff
@@ -383,6 +383,38 @@ vector<UsingForDirective const*> usingForDirectivesForType(Type const& _type, AS
 
+set<FunctionDefinition const*, ASTNode::CompareByID> Type::operatorDefinitions(
+	Token _token,
+	ASTNode const& _scope,
```
Shouldn't you just always be scanning the scope of ``typeDefinition()``?

cameel replied:
Now that operators are global, probably. It used to be different when they could be defined in contracts and libraries.

I thought about simplifying that part, but in the end I decided to keep it as is for various reasons. It works, there was still a small chance that operators would not be global and, finally, I will change it soon anyway. After the PR that stores operator definitions in a `SourceUnit` annotation, this function will just be returning stuff from that annotation.

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -775,6 +799,31 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(*function->returnParameters()[0]->type() == *_binOp.annotation().type);
+
+		string left = expressionAsType(_binOp.leftExpression(), *function->parameters()[0]->type());
+		string right = expressionAsType(_binOp.rightExpression(), *function->parameters()[1]->type());
```
```suggestion
		string left = expressionAsCleanedType(_binOp.leftExpression(), *function->parameters()[0]->type());
		string right = expressionAsCleanedType(_binOp.rightExpression(), *function->parameters()[1]->type());
```

Or why don't we need cleanup here?

cameel replied:
Not sure. I will need to check. I was assuming that the cleanup happens somewhere inside `expressionAsType()`. It did not break any tests, but to be honest, I don't have tests for cleanup in custom operators (do we even have such tests for built-in ones?) so it could actually be broken.

ekpyron replied:
Yeah, this is generally done a bit confusingly, since in the middle of building this originally we switched strategies between assuming stuff being always clean and always cleaning stuff explicitly (which can make a difference in the presence of inline assembly only, but there it can...)...
Anyways, since the comparison operators do this, we should do it here as well. I'd guess the following would yield different results depending on it:

```solidity
type T is uint8;
function f(T x, T y) returns (T z) {
  assembly { z := shr(8, x) }
}
using {f as +} for T;
contract C {
  function test() external returns (T) {
    T a;
    assembly { a := 0x4200 }
    return a + a;
  }
}

```

Probably no way to trigger a difference without inline assembly on both sides :-).

cameel replied:
After looking at this, I think the current behavior (i.e. no cleanup) is correct because it matches what happens when you call the function directly. In your test, if you replace `a + a` with `f(a, a)`, you get 42, not 0.

Also, I think that the whether we do cleanup for built-in operators or not does not matter here because no one can really observe whether it happens or not. In either case the implementation provided by the compiler will account for it and calculate the result correctly. It may or may not clean the result (I did not check), but even then I think the behavior of user-defined operators should be the same as when the function is called directly and does not necessarily have to match the behavior of the built-in operator.

I added tests (`calling_operator_parameter_cleanup.sol` and `calling_operator_return_parameter_cleanup.sol`) but due to this I put in the expectation that the parameters and return values are **not** cleaned.

ekpyron replied:
I mean, you have a test case right there that observes it, so it can be observed :-).
The part with function calls is some argument, but arguably we could even consider that a bug...
It's a bit tricky due to that, but I'd *still* do the cleanup actually... let me think...

ekpyron replied:
It's not about matching the behaviour of builtin operators, it's about avoiding weird behaviour - this is especially important since operator definitions are quite likely to be done in assembly.
```
type T is uint8;
function f(T x, T y) returns (bool z) {
  assembly { z := eq(x, y) }
}
using {f as ==} for T;
contract C {
  function test() external returns (bool) {
    T a; T b = T.wrap(0);
    assembly { a := 0x4200 }
    return a == b;
  }
}
```
This *should* return true...

ekpyron replied:
But yeah, a bit unfortunate that we don't do it for internal function calls, which makes this inconsistent no matter what...

chriseth replied:
I think it's consistent to do the same thing we do for internal function calls. It is just syntactic sugar after all.

chriseth replied:
if you implement a custom comparison operator in inline assembly, you have to think about what to do with unclean inputs. This way we actually leave it open to the implementer what to do.

ekpyron replied:
Well, @chriseth is also for keeping it without cleanup, so let's keep it as is.

ekpyron replied:
> if you implement a custom comparison operator in inline assembly, you have to think about what to do with unclean inputs. This way we actually leave it open to the implementer what to do.

Not sure if this isn't a bit reckless... I don't expect people who write implementations for these will take unclean input into account - I mean we missed that a few times ourselves in the past...

But we can hope that small types as underlying types is uncommon anyways and should properly rethink the current cleanup mess in 0.9.

cameel replied:
> I mean, you have a test case right there that observes it, so it can be observed :-).

This case observes the behavior of a user-defined operator. How can you observe that for a built-in one? You cannot really inject assembly into it.

cameel replied:
> This should return true...

Why though? It explicitly bypasses the high-level mechanisms to accesses the whole slots containing `x` and `y` rather than just their `uint8` part. I'm actually rather surprised that we even want to guarantee this level of cleanup.

And, actually, why would it ever be ok to clean up for an operator and not a function? You can do the exact same mistake when calling a function that you implemented in assembly.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -502,7 +530,34 @@ bool ExpressionCompiler::visit(BinaryOperation const& _binaryOperation)
+		solAssert(functionType->kind() == FunctionType::Kind::Internal);
+
+		evmasm::AssemblyItem returnLabel = m_context.pushNewTag();
+		acceptAndConvert(leftExpression, *functionType->parameterTypes()[0]);
+		acceptAndConvert(rightExpression, *functionType->parameterTypes()[1]);
```
```suggestion
		acceptAndConvert(leftExpression, *functionType->parameterTypes()[0], true);
		acceptAndConvert(rightExpression, *functionType->parameterTypes()[1], true);
```

cameel replied:
Left as is. See https://github.com/ethereum/solidity/pull/13790#discussion_r1113626349.

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -672,6 +672,30 @@ void IRGeneratorForStatements::endVisit(Return const& _return)
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(*function->returnParameters()[0]->type() == *_unaryOperation.annotation().type);
+
+		string argument = expressionAsType(_unaryOperation.subExpression(), *function->parameters()[0]->type());
```
```suggestion
		string argument = expressionAsCleanedType(_unaryOperation.subExpression(), *function->parameters()[0]->type());
```

cameel replied:
Left as is. See https://github.com/ethereum/solidity/pull/13790#discussion_r1113626349.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -410,6 +410,34 @@ bool ExpressionCompiler::visit(TupleExpression const& _tuple)
+		solAssert(functionType->kind() == FunctionType::Kind::Internal);
+
+		evmasm::AssemblyItem returnLabel = m_context.pushNewTag();
+		acceptAndConvert(_unaryOperation.subExpression(), *functionType->parameterTypes()[0]);
```
```suggestion
		acceptAndConvert(_unaryOperation.subExpression(), *functionType->parameterTypes()[0], true);
```

cameel replied:
Left as is. See https://github.com/ethereum/solidity/pull/13790#discussion_r1113626349.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1729,31 +1730,67 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
+	);
+
+	// Operator can't be both user-defined and built-in at the same time.
+	solAssert(!builtinResult || matchingDefinitions.empty());
```
But they have to be *either* user-defined *or* built-in, don't they :-)?
I.e.

```suggestion
	solAssert(!builtinResult != matchingDefinitions.empty());
```
would be the better assertion, wouldn't it :-)?

ekpyron replied:
Yeah, that's an explicit error case below... might be easier to read to put this closer together with that below, but also fine.

cameel replied:
Well, it's pretty much next to the `if`. I could put the `else` case first but not sure it actually improves things. It then moves other things away :)

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3892,6 +3993,145 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
+				);
+			}
+
+			TypePointers const& parameterTypes = functionType->parameterTypesIncludingSelf();
```
``functionType`` is just the type of the function definition, right? So there is nothing bound in that and there's no difference between ``parameterTypes`` and ``parameterTypesIncludingSelf``?

cameel replied:
Yes. Added an assertion that it's not bound and changed to `parameterTypes()`.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3892,6 +3993,145 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
+			bool identicalFirstTwoParameters = (parameterCount < 2 || *parameterTypes.at(0) == *parameterTypes.at(1));
+			bool isUnaryOnlyOperator = (!TokenTraits::isBinaryOp(operator_.value()) && TokenTraits::isUnaryOp(operator_.value()));
+			bool isBinaryOnlyOperator =
+				(TokenTraits::isBinaryOp(operator_.value()) && !TokenTraits::isUnaryOp(operator_.value())) ||
+				operator_.value() == Token::Add;
```
This is rather weird :-). I see that ``isUnaryOp`` still has ``Token::Add`` as a ``true``-case... we should really change that, that's probably an artifact of ancient times only :-). But we can (and should) do it separately (if we don't forget).

cameel replied:
True, I'll add it to my list of things to clean up afterwards.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3892,6 +3993,145 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
+		else if (operator_.has_value())
+		{
+			if (!_usingFor.global())
+			{
+				m_errorReporter.typeError(
+					3320_error,
+					path->location(),
+					"Operators can only be defined in a global 'using for' directive."
+				);
+			}
```
```suggestion
				m_errorReporter.typeError(
					3320_error,
					path->location(),
					"Operators can only be defined in a global 'using for' directive."
				);
```

cameel commented:
I think I fixed all there was to fix. The commits are squashed. Here's a rough list of changes:
- Extra semantic tests for cleanup.
- Extra semantic test for an operator calling a `view` function via a mismatched interface.
- An error about the type not being an UDVT does not hide an error about the function not being free or pure (+test case).
- Extra assert about function not being bound when type-checking `using for`.
- Changed naming convention for semantic tests. The `calling_operator_` prefix was not helpful as it is in syntax tests, because all semantic tests necessarily invoke the operator.
